### PR TITLE
feat(mutation): can mutate lambdas

### PR DESCRIPTION
**Pull request by arnobl**
First changes for #123

It adds a stopper 'lambda' not used by default.
For example: `+STOP_METHODS(except[lambda])` allows the mutation of lambdas.

Works like a charm on https://github.com/interacto/interacto-java-api

Regarding the changes, I am not a big fan of 
```java
static StopMethodMatcher isSynthetic() {
	return (classTree, methodTree) -> methodTree.isSynthetic() && !methodTree.rawNode().name.startsWith("lambda$");
}

static StopMethodMatcher isLambda() {
	return (classTree, methodTree) -> methodTree.isSynthetic() && methodTree.rawNode().name.startsWith("lambda$");
}
```
as it contains some duplicate stuffs. Cannot find a way to combine matchers.
This is required to specify in `isSynthetic` that lambdas must be ignored as otherwise the stopper 'synthetic' will also filter out lambdas.

I also had to change `DelegateMethodMatcherTest.java`.

### CODE REVIEW

**Comment by arnobl** [comment]
An small issue I identified:
```
1. lambda$isGuardOK$0 : All method instructions replaced by: return true → SURVIVED
2. isGuardOK : All method instructions replaced by: return false → KILLED
3. isGuardOK : All method instructions replaced by: return true → KILLED
4. lambda$isGuardOK$0 : All method instructions replaced by: return false → KILLED
```

with the code:
```java
	@Override
	protected boolean isGuardOK(final E event) {
		return findTransition(event)
			.filter(tr -> tr.isGuardOK(event))
			.isPresent();
	}
```
It seems that descartes now creates redundant mutants.

**Comment by oscarlvp** [comment]
Mutant 1:
```java
protected boolean isGuardOK(final E event) {
    return findTransition(event)
        .filter(tr -> true)
        .isPresent();
}
```
Mutant 2:
```java
protected boolean isGuardOK(final E event) {
    return false;
}
```
Mutant 3:
```java
protected boolean isGuardOK(final E event) {
    return true;
}
```
Mutant 4
```java
protected boolean isGuardOK(final E event) {
    return findTransition(event)
        .filter(tr -> false)
        .isPresent();
}
```

So, 1 and 4 are for the lambda, 2 and 3 for the method. There is a difference.
So in this case, it seems that, in this test code, the lambda invocation always returns `true`.

**Comment by oscarlvp** [comment]
One, question, why do we have to consider lambdas as stop methods?
What we can do is to simply remove lambdas from the `isSynthetic`. We can mutate all lambdas, there is no issue with that.

File: src/main/java/eu/stamp_project/mutationtest/descartes/interceptors/stopmethods/StopMethodMatchers.java [hunk 1]
```diff
@@ -45,9 +45,13 @@ static StopMethodMatcher isEmptyVoid() {
...
     }
 
-    static StopMethodMatcher isSynthetic() {
-        return (classTree, methodTree) -> methodTree.isSynthetic();
-    }
+	static StopMethodMatcher isSynthetic() {
+		return (classTree, methodTree) -> methodTree.isSynthetic() && !methodTree.rawNode().name.startsWith("lambda$");
+	}
+
+	static StopMethodMatcher isLambda() {
```
Comment (oscarlvp) [inline src/main/java/eu/stamp_project/mutationtest/descartes/interceptors/stopmethods/StopMethodMatchers.java]: I think there is no need to create this `StopMethodMatcher` just the exclusion of lambdas in the `isSynthetic` matcher would be OK. Simple lambdas will be filtered by other matchers.

**Comment by arnobl** [comment]
> One, question, why do we have to consider lambdas as stop methods?
What we can do is to simply remove lambdas from the isSynthetic. We can mutate all lambdas, there is no issue with that.

you're the boss. I though it could be an option. But if you think this should be included by default, I'm ok.

File: src/main/java/eu/stamp_project/mutationtest/descartes/interceptors/stopmethods/StopMethodMatchers.java (hunk 1 above)
Comment (arnobl) [reply src/main/java/eu/stamp_project/mutationtest/descartes/interceptors/stopmethods/StopMethodMatchers.java]: ok. Done. Lambdas are now mutated by default (no more stopper).

**Pull request merged**